src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -146,12 +155,73 @@ class GcsTaskManager : public rpc::TaskInfoHandler {
     std::vector<rpc::TaskEvents> GetTaskEvents(
         const absl::flat_hash_set<TaskAttempt> &task_attempts) const;
 
+    /// Mark the task tree containing this task attempt as failure if necessary.
+    ///
+    /// A task tree is represented by the `parent_task_id` field for each task.
+    ///
+    /// All non-terminated (not finished nor failed) tasks in the task tree with the task
+    /// event as the root should be marked as failed with the root's failed timestamp if
+    /// any of:
+    ///     1. Its parent has failed: this happens when a task event is reported from a
+    ///     worker later than the failure status reported from the parent worker.
+    ///     2. The task itself fails.
+    ///
+    /// This is necessary since workers/drivers/raylet could crash, and terminal task
+    /// status are not reported when that happens. Here we leverages on Ray's fault
+    /// tolerance where a parent task failure will lead to all child task fail to mark the
+    /// task tree with failure status recursively.
+    ///
+    /// \param root_task_id ID of the task event that's the root of the task tree.
+    /// \param parent_task_id ID of the task's parent.
+    void MarkTaskTreeFailed(const TaskID &root_task_id, const TaskID &parent_task_id);
+
+    /// Update a task event.
+    ///
+    /// \param task_attempt The task attempt (task id + attempt number).
+    /// \param callback The update callback that updates the task event.
+    /// \return True if the update is successful.
+    bool UpdateTaskEvent(const TaskAttempt &task_attempt,
+                         const UpdateTaskEventCallback &callback);
```

Comment:
There wasn't an API that returns a reference or pointer to the underlying data so I thought using callback would be better, and might potentially work with future iterations of decoupling the storage. But yeah, this might have been over-engineering. I think maybe adding another function that returns mutable reference is less confusing.